### ISSUE TITLE
repeat: -1

### ISSUE THREAD

**Issue by tobireif** [legomushroom/mojs#78]
Does MoJS support "repeat: [infinity]" ? Syntax could be repeat: -1.

**Comment by legomushroom** [comment]
Hi, Tobi! Thanks for the issue!

Right now it doesn't, and because of some architecture constraints it kind of hard to implement. We will probably add some `public` method for that soon, probably something like `playInfinite` or `replayInfinite` will do. At the moment you can just call `replay` in `onComplete` callback:

```javascript
const shape = new mojs.Shape({
  scale: { 0: 1 },
  duration: 2000,
  onComplete: function () { this.replay(); } // <------
}).play();
```
There is [codepen demo](http://codepen.io/sol0mka/pen/BQdyVW?editors=0010#). Hope it helps!

**Comment by tobireif** [comment]
Hi Oleg, thanks for your reply!

Your suggestion ~ "replayInfinitely" would be much more expressive / easier to read than my suggestion / GSAP's syntax "-1".

**Comment by harrynewsome** [comment]
Am I right in saying you will have different setting for `playInfinite|replayInfinite` rather than having `repeat: foo`? 

If so what about `loop`.

```javascript
const shape = new mojs.Shape({
  scale: { 0: 1 },
  duration: 2000,
  loop: true, // <------
}).play();
```

**Comment by legomushroom** [comment]
Hi, Tobi! Great, we should add it then!

// @tobireif

**Comment by legomushroom** [comment]
Hi, Harry!

Hm, `loop` property will work great too. We should call it `isLoop` to be consistent with other `boolean` properties. So I don't know, we can add both the `isLoop` and `playInfinite|replayInfinite` options `or` need to pick just one of them now. What would you guys prefer more?

// @harrynewsome @tobireif

**Comment by legomushroom** [comment]
*Just a few thoughts on `repeat: -1|Infinity`*

In `mojs` everything is timeline-based. So your animation is "placed" on some sort of timeline and you can move a sort of "control handle" to set animation's progress regarding the time - just like with [mojs-player](https://github.com/legomushroom/mojs-player) or with `setProgress` method. The `repeat` property defines how many times the animation is being "copied" on that timeline, consider the next example with `repeat: 2`:

```
    original animation          repeat 1               repeat 2
[----------------------][----------------------][----------------------]

Animation with `repeat: 2` plays 3 times - the original period + 1st repeat period + 2nd repeat period.
```
So in this system we have 2 contradictions:
- `repeat: -1` does not make any sense. repeat period defines how the `origin period` is expanded on the timeline and having `-1` rather mean collapsing the `original period` instead
- `repeat: Infinity` makes more sense but it means that we need to expand the animation to `infinite` number of periods that will break the `setProgress` method. There is no sense to set `setProgress(.5)` with `infinite` number of periods -- what is the `.5`(half) of `Infinity`? What period will it be in then? It is a `non-deterministic` value so we can't know.
Secondly, if we have `infinite` animation, timeline tools like [mojs-player](https://github.com/legomushroom/mojs-player) will be broken for the same reasons as the `setProgress` is - if you will try to move the player's handle around - it will be stuck at the same position because entire player's track represents infinity now, and if you move 2 seconds right or left in the infinity - you make no progress at all, - resolution of the track is too low - `infinity`.

But it makes total sense to have `isLoop` or `playInifinte` instead. It will mean that we have only one period of the animation on the timeline - just the `original period`. And when it stops playing - we simply launch it again from the beginning over and over. So the `semantics` of the `APIs` will imply that user should not expect `infinite` number of periods on the timeline, but rather automatic `relaunch loop` when animation stops.

**Comment by tobireif** [comment]
> repeat: -1 does not make any sense.

True! It would be a special number of the API which would mean "repeat: indefinitely". But it would not be very expressive / clear, so having a nice word instead of a "special number" wold be much better.

```isLoop: true``` seems fine to me (alternative name eg ```repeatIndefinitely: true```), or eg ```repeat: "indefinitely"``` (a string value meaning "loop this").

**Comment by Sandstedt** [comment]
I vote for isLoop, much easier to spell 😊. 
Then we have After Effects loopOut() expression. Maybe an alternative? ... Nah, go with isLoop.

**Comment by sdras** [comment]
I vote for isLoop, too, it's concise and fits in with the pattern of the rest of the API

**Comment by legomushroom** [comment]
Looks like the isLoop is total winner, thank you guys for all the feedback!

**Comment by tobireif** [comment]
Te advantage of ```repeat:"indefinitely"``` would be that the strange combination of eg ```repeat:5``` *and* ```isLoop:true``` wouldn't be possible. With ```repeat:"indefinitely"```, the API user could only have eg ```repeat:5``` *or* ```repeat:"indefinitely"```, less ambiguous. But ```isLoop``` would be OK enough - when the API user has ```repeat:5``` *and* ```isLoop:true```, the latter probably should win.

**Comment by harrynewsome** [comment]
@tobireif I see what you mean, I suggested it because it was already an idea to have another setting for the looping. `repeat: true` seems like a good fit too :)

thoughts @legomushroom?

**Comment by tobireif** [comment]
@harrynewsome Nice one! I can't imagine a better option than ```repeat: true```.

**Comment by nwalton3** [comment]
+1 for `repeat: true`

**Comment by Sandstedt** [comment]
But if loop is a separate method, you could use, say a tween, with repeat: 3 with a repeat delay of 10s, and also isLoop: true. Then we will copy then tween 3 times into the timeline, then wait 10s and repeat 3 times again into infinity.

**Comment by legomushroom** [comment]
Sorry for the delay guys! @tobireif @harrynewsome @nwalton3 @Sandstedt 

It should be separate property - this way it will be slightly more flexible for the reasons listed in the previous comment - let's say you want to have some sort of spinner that goes back and forth in a loop. So you probably want to make just two periods for the animation - one for forward direction and for backward direction. Also, if you use tools like `mojs-player` or upcoming `mojs-timeline-editor` - you want to debug exactly those two periods inside them. When you are done debugging - you set the `isLoop` property and hit `play` - you will have your two periods repeating infinitely.

Another reason to have them separately I've listed in the `Just a few thoughts on repeat: -1|Infinity` comment at the top. In short - `Infinity` is broken - it implies that you have just one animation period that you want to `loop`, and it is `none-deterministic` - you cannot set `50%` progress of infinity - so you can't use `setProgress` method and debugging tools.

So I would rather stick with separate properties:
- `repeat`: `Integer >= 0` - the property that defines how many repeat periods the animation will have - this is how the animation will be shown in debugging tools. These repeat periods are, of course, are affected by `isYoyo` property.
- `isLoop`: `Boolean` - the property defines that we should run animation again when it is done `playing` - so we have to hit `.play()` to use this property - otherwise it will do nothing. SO this property only for `playback` control. This property is not affected by `isYoyo` one.

**Comment by harrynewsome** [comment]
Thanks for explaining, `isLoop` sounds good :)

// @Sandstedt @legomushroom

**Comment by tobireif** [comment]
@legomushroom All up to you :) Thanks for Mo.js!

**Comment by nwalton3** [comment]
@legomushroom Sounds great. Thanks for explaining.
